src/librustc_typeck/diagnostics.rs

```diff
@@ -3426,6 +3426,30 @@ parameters. You can read more about it in the API documentation:
 https://doc.rust-lang.org/std/marker/struct.PhantomData.html
 "##,
 
+E0393: r##"
+A type parameter which needs to be explicity specified because its
+default value is `Self` isn't. Example of erroneous code:
+
+```compile_fail
+trait A<T=Self> {}
+
+fn together_we_will_rule_the_galaxy(son: &A) {}
+// error: the type parameter `T` must be explicitly specified in an
+          object type because its default value `Self` references the
+          type `Self`
+```
+
+Since `Self` represents the type on which the trait is implemented,
```

Comment:
No, it doesn't explain why that's necessary in this case. Default type params work otherwise.

This might work:

A trait object is defined over a single, fully-defined trait. With a regular default parameter, this parameter can just be substituted in. However, if the default parameter is `Self`, the trait changes for each concrete type; i.e. `i32` will be expected to implement `A<i32>`, `bool` will be expected to implement `A<bool>`, etc. These types will not share an implementation of a fully-defined trait; instead they share implementations of a trait with different parameters substituted in for each implementation. This is irreconcilable with what we need to make a trait object work, and is thus disallowed. Making the trait concrete by explicitly specifying the value of the defaulted parameter will fix this issue.